interpreter/functions/formatting.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functions
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// FormatCallback is an interface that allows user-defined behavior
+// for formatting clause implementations, as well as argument retrieval.
+type FormatCallback interface {
+	String(ref.Val, string) (string, error)
+	Decimal(ref.Val, string) (string, error)
+	Fixed(*int) func(ref.Val, string) (string, error)
+	Scientific(*int) func(ref.Val, string) (string, error)
+	Binary(ref.Val, string) (string, error)
+	Hex(bool) func(ref.Val, string) (string, error)
+	Octal(ref.Val, string) (string, error)
+	Arg(int64) (ref.Val, error)
+	ArgSize() int64
```

Comment:
These are not like the others. What do they do?

Reply:
And should these be in the same interface? All the format directive handler functions clearly belong together in an interface. But `Arg` and `ArgSize` seem to serve a very different purpose.

Reply:
It's used during parsing (`ArgSize` here: https://github.com/google/cel-go/pull/645/files#diff-3eaeae7b0bf9e5fba3c8585d57a53a450b0476fedc18fefece60812d54a38c66R90-R92 and `Arg` here: https://github.com/google/cel-go/pull/645/files#diff-3eaeae7b0bf9e5fba3c8585d57a53a450b0476fedc18fefece60812d54a38c66R101) - I agree a separate interface would probably be better (call it `FormatList` or something).

Reply:
Separate interface + more documentation SGTM

Reply:
Gotcha, fixed.